File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head):
        prev = None
        while head:
            curr = head
            head = head.next
            curr.next = prev
            prev = curr
        return prev
    
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        pnodeleft = dummy = ListNode(-1, head)
        nodeleft = head
        
        for _ in range(left-1):
            pnodeleft = nodeleft
            nodeleft = nodeleft.next
            
        pnoderight = None
        noderight = head
        
        for _ in range(right):
            pnoderight = noderight
            noderight = noderight.next
        
        pnodeleft.next = None
        pnoderight.next = None
        
        nodeleft = self.reverseList(nodeleft)
        
        pnodeleft.next = nodeleft
        
        while nodeleft and nodeleft.next:
            nodeleft = nodeleft.next
            
        nodeleft.next = noderight

        return dummy.next
            
            

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pnodeLeft = dummy;
        ListNode lnode = head;
        
        while(left>1){
            pnodeLeft = lnode;
            lnode = lnode.next;
            left--;
            right--;
        }
        
        ListNode pnodeRight = lnode;
        ListNode rnode = lnode.next;
        
        while(right>1){
            pnodeRight = rnode;
            rnode = rnode.next;
            right--;
        }
        
        pnodeLeft.next = null;
        pnodeRight.next = null;
        
        lnode = reverse(lnode);
        
        pnodeLeft.next = lnode;
        
        while(lnode!=null && lnode.next!=null){
            lnode = lnode.next;
        }
        
        lnode.next = rnode;
        
        return dummy.next;
    }
    
    ListNode reverse(ListNode head){
        ListNode prev = null;
        while(head!=null){
            ListNode temp = head.next;
            head.next = prev;
            
            prev = head;
            head = temp;
        }
        return prev;
    }
}